ambry-rest/src/test/java/com/github/ambry/rest/NettyResponseChannelTest.java

```diff
@@ -1082,6 +1084,33 @@ private void sendRequestAndEvaluateResponsePerformance(HttpRequest httpRequest,
       assertFalse(httpRequest.method() + " request should be unsatisfied", MockNettyRequest.mockTracker.isSatisfied());
     }
   }
+
+  /**
+   * Tests the invocation of DELAYED_CLOSE when post failures happen in {@link NettyResponseChannel}.
+   */
+  @Test
+  public void completeRequestWithDelayedCloseTest() throws Exception {
+    Properties properties = new Properties();
+    properties.setProperty(NettyConfig.NETTY_SERVER_CLOSE_DELAY_TIMEOUT_MS, "500");
+    NettyConfig nettyConfig  = new NettyConfig(new VerifiableProperties(properties));
+    MockNettyMessageProcessor processor = new MockNettyMessageProcessor();
+    processor.setNettyConfig(nettyConfig);
+    ChunkedWriteHandler chunkedWriteHandler = new ChunkedWriteHandler();
+    EmbeddedChannel channel = new EmbeddedChannel(chunkedWriteHandler, processor);
+
+    RestServiceErrorCode REST_ERROR_CODE = RestServiceErrorCode.BadRequest;
+    String content = "@@randomContent@@@";
+    HttpHeaders httpHeaders = new DefaultHttpHeaders();
+    httpHeaders.set(MockNettyMessageProcessor.REST_SERVICE_ERROR_CODE_HEADER_NAME, REST_ERROR_CODE);
+    httpHeaders.set(MockNettyMessageProcessor.INCLUDE_EXCEPTION_MESSAGE_IN_RESPONSE_HEADER_NAME, "true");
+    HttpRequest httpRequest = RestTestUtils.createFullRequest(HttpMethod.POST,
+        TestingUri.OnResponseCompleteWithRestException.toString(), httpHeaders, content.getBytes());
+    channel.writeInbound(httpRequest);
+    HttpResponse response = (HttpResponse) channel.readOutbound();
+    assertEquals("Unexpected response status", getExpectedHttpResponseStatus(REST_ERROR_CODE), response.status());
+    //channel should not be closed right away.
+    assertTrue("Channel closed on the server", channel.isActive());
```

Comment:
Can also wait 500ms and verify that it does get closed.  Also check the counters were incremented.

Reply:
I actually tried waiting but it did not work. There is no dedicated threads running the netty channel handlers here.  Everything is getting executed in a single Junit test runner thread. The netty eventloop.scheduler() task never gets executed because netty does not have any real threads to schedule this task. Even if i add a Thread.sleep() here, after the sleep is over the the test runner thread simply exits. I could not find any other way of verifying this. Let me know if there is any way to test this. 

I added check for the metrics. This is fine.

Reply:
I'm kind of surprised by this.  Looking at the code, I see EmbeddedChannel uses EmbeddedEventLoop which extends AbstractScheduledEventExecutor and I would have thought the runScheduledTasks() method would get the scheduled task and run it.  Did you try stepping through with the debugger to see where the task ends up?  I see that Yingyi @jsjtzyy has worked on this test class, he might have suggestions on how to trigger it.

Reply:
Thanks so much @lightningrob for the pointers. Since EmbeddedChannel does not have its own thread pool the UT thread has to call the runPendingTask() api to fire the scheduled tasks on this channel. This is very different from the NioSocketChannel. I was not aware that the api has to be explicitly triggered. I have enhanced the UT to do all the validations now. Please take a look. Thanks.